crates/viewer/re_blueprint_tree/src/data.rs

```diff
@@ -0,0 +1,515 @@
+                // in the projection (or shouldn't be included if the projection has already
+                // been filtered out).
+                false
+            }
+        });
+
+        let projection_trees = projections
+            .into_iter()
+            .filter_map(|node| {
+                let projection_tree = DataResultData::from_data_result_and_filter(
+                    ctx,
+                    view_blueprint,
+                    query_result,
+                    &DataResultNodeOrPath::DataResultNode(node),
+                    true,
+                    &mut hierarchy,
+                    filter_matcher,
+                    false,
+                );
+
+                if cfg!(debug_assertions) {
+                    debug_assert!(hierarchy.is_empty());
+                } else {
+                    hierarchy.clear();
+                }
+
+                projection_tree
+            })
+            .collect_vec();
+
+        if origin_tree.is_none() && projection_trees.is_empty() {
+            return None;
+        }
+
+        let default_open = filter_matcher.is_active()
+            || origin_tree.as_ref().map_or(true, |data_result_data| {
+                default_open_for_data_result(data_result_data.children.len())
+            });
+
+        Some(Self {
+            id: view_blueprint.id,
+            class_identifier: view_blueprint.class_identifier(),
+            name: view_blueprint.display_name_or_default(),
+            visible: view_blueprint.visible,
+            default_open,
+            origin_tree,
+            projection_trees,
+        })
+    }
+}
+
+// ---
+
+/// The various kind of things that may be represented in a data result tree.
+#[derive(Debug)]
+#[cfg_attr(feature = "testing", derive(serde::Serialize, serde::Deserialize))]
+pub enum DataResultKind {
+    /// This is a regular entity part of a data result (or the tree that contains it).
+    EntityPart,
+
+    /// When the view has no data result contained in the view origin subtree, we still display the
+    /// origin with a warning styling to highlight what is likely an undesired view configuration.
+    EmptyOriginPlaceholder,
+
+    /// Since the view origin tree is displayed on its own, we don't repeat it within projections
+    /// and instead replace it with this placeholder.
+    OriginProjectionPlaceholder,
+}
+
+#[derive(Debug)]
+#[cfg_attr(feature = "testing", derive(serde::Serialize, serde::Deserialize))]
+pub struct DataResultData {
+    pub kind: DataResultKind,
+    pub entity_path: EntityPath,
+    pub visible: bool,
+
+    pub view_id: ViewId,
+
+    /// Label that should be used for display.
+    ///
+    /// This typically corresponds to `entity_path.last().ui_string()` but covers corner cases.
+    pub label: String,
+
+    /// The sections within the label that correspond to a filter match and should thus be
+    /// highlighted.
+    pub highlight_sections: SmallVec<[Range<usize>; 2]>,
```

Comment:
Using `2` here is surprising. I would expect you would most often have either zero or one highlights?

```suggestion
    pub highlight_sections: SmallVec<[Range<usize>; 1]>,
```

Reply:
Makes sense!